src/crypto/mod.rs

```diff
@@ -108,6 +110,17 @@ impl<E: Engine> PartialEq for SecretKey<E> {
     }
 }
 
+impl<E: Engine> Drop for SecretKey<E> {
+    fn drop(&mut self) {
+        let size = mem::size_of_val(self);
+        unsafe {
+            let p = self as *mut Self;
+            ptr::write_bytes(p as *mut u8, 0, size);
+            ptr::write(p, SecretKey(E::Fr::zero()));
```

Comment:
In the `Bls12` implementation this is indeed represented by all zeros:
https://github.com/ebfull/pairing/blob/09b6e6f9212020f385218e5cf5287e381ccd312b/src/bls12_381/fr.rs#L327
What if it is implemented differently for another curve? Maybe the `Drop` instance should be specified only for `SecretKey<Bls12>` since this is the only one where this instuction is guaranteed to work as we expect?

Or maybe the `write` instruction is not needed after all?

Reply:
The call to `write` was meant to re-instantiate a valid SecretKey object, I used `zero()` just because it was the easiest way to create a new valid object. The `zero()` call wasn't meant to zero-out memory (that's what `write_bytes()` is doing).

I thought that re-instantiating was necessary because after `SecretKey::drop()` is called,  Rust will call `std::mem::drop()` behind the scenes, which takes ownership of an argument `T` (in our case `T` is a `SecretKey`), and then frees the memory of `T`.

But, I just tried the above tests with `write()` commented out, and everything builds and passes. So, I think that you are correct, in that the call to `write()` is unnecessary.

Reply:
It's not really important that it's all zeros, is it? As long as every byte is overwritten with anything, we know that we can't leak a private key.

But it sounds like [clear_on_drop](https://crates.io/crates/clear_on_drop) would be the safest choice.

Reply:
My understanding is that what is also important with regards to preventing the leakage or secret keys is avoiding moving variables. Every time a variable is moved it's copied. Hopefully the upcoming Pin/Unpin thing will be of use for this case https://github.com/rust-lang/rfcs/blob/master/text/2349-pin.md